Expose script flags to processCommand for better handling

oranagra opened the pull request:
In order to better handle scripts with shebang flags that can indicate
the script is a read-only, allow-oom, allow-stale etc, or not.

The important part is that read-only scripts (not just EVAL_RO
and FCALL_RO, but also ones with `no-writes` executed by normal EVAL or
FCALL), will now be permitted to run during CLIENT PAUSE WRITE (unlike
before where only the _RO commands would be processed).

Other than that, some errors like OOM, READONLY, MASTERDOWN are now
handled by processCommand, rather than the command itself affects the
error string (and even error code in some cases), and command stats.

Besides that, now the `may-replicate` commands, PFCOUNT and PUBLISH, will
be considered `write` commands in scripts and will be blocked in all
read-only scripts just like other write commands.
They'll also be blocked in EVAL_RO (i.e. even for scripts without the
`no-writes` shebang flag.

This commit also hides the `may_replicate` flag from the COMMAND command
output. this is a **breaking change**.

background about may_replicate:
We don't want to expose a no-may-replicate flag or alike to scripts, since we
consider the may-replicate thing an internal concern of redis, that we may
some day get rid of.
In fact, the may-replicate flag was initially introduced to flag EVAL: since
we didn't know what it's gonna do ahead of execution, before function-flags
existed). PUBLISH and PFCOUNT, both of which because they have side effects
which may some day be fixed differently.

code changes:
The changes in eval.c are mostly code re-ordering:
- evalCalcFunctionName is extracted out of evalGenericCommand
- evalExtractShebangFlags is extracted luaCreateFunction
- evalGetCommandFlags is new code

This was discussed in https://github.com/redis/redis/pull/10364#issuecomment-1125119110

Doc PR: https://github.com/redis/redis-doc/pull/1954

oranagra commented on tests/unit/functions.tcl:
```diff
@@ -424,7 +424,7 @@ start_server {tags {"scripting repl external:skip"}} {
             } e
             set _ $e
-        } {*Can not run script with write flag on readonly replica*}
+        } {READONLY You can't write against a read only replica.}
```
used to be `ERR Can not run script with write flag on readonly replica`

oranagra commented on tests/unit/functions.tcl:
```diff
@@ -1052,7 +1052,7 @@ start_server {tags {"scripting"}} {
 
         catch {[r fcall f1 1 k]} e
-        assert_match {*can not run it when used memory > 'maxmemory'*} $e
+        assert_match {OOM *when used memory > 'maxmemory'*} $e
```
used to be `OOM allow-oom flag is not set on the script, can not run it when used memory > 'maxmemory`

oranagra commented on tests/unit/functions.tcl:
```diff
@@ -1085,12 +1085,12 @@ start_server {tags {"scripting"}} {
 
         catch {[r fcall f1 0]} e
-        assert_match {*'allow-stale' flag is not set on the script*} $e
+        assert_match {MASTERDOWN *} $e
```
was already `MASTERDOWN` but the one in script.c

oranagra commented on tests/unit/functions.tcl:
```diff
@@ -1085,12 +1085,12 @@ start_server {tags {"scripting"}} {
 
         catch {[r fcall f3 0]} e
-        assert_match {*Can not execute the command on a stale replica*} $e
+        assert_match {ERR *Can not execute the command on a stale replica*} $e
```
The error message didn't change in this PR (just the test)

oranagra commented on tests/unit/scripting.tcl:
```diff
@@ -1416,7 +1416,7 @@ start_server {tags {"scripting"}} {
 
         # Fail to execute regardless of script content when we use default flags in OOM condition
-        assert_error {OOM allow-oom flag is not set on the script, can not run it when used memory > 'maxmemory'} {
+        assert_error {OOM *} {
```
now returns a normal OOM error

oranagra commented on tests/unit/scripting.tcl:
```diff
@@ -1458,11 +1458,53 @@ start_server {tags {"scripting"}} {
             ] "some value"
 
-            assert_error {ERR Can not run script with write flag on readonly replica} {
+            assert_error {READONLY You can't write against a read only replica.} {
```
now return a normal READONLY error

oranagra commented on tests/unit/scripting.tcl:
```diff
@@ -1476,7 +1518,7 @@ start_server {tags {"scripting"}} {
         }
 
-        assert_error {*'allow-stale' flag is not set on the script*} {
+        assert_error {MASTERDOWN Link with MASTER is down and replica-serve-stale-data is set to 'no'.} {
```
now returns a normal MASTERDOWN error

oranagra commented on src/script.c:
```diff
@@ -118,18 +129,18 @@ int scriptPrepareForRun(scriptRunCtx *run_ctx, client *engine_client, client *ca
         }
 
-        if (!(script_flags & SCRIPT_FLAG_ALLOW_OOM) && server.script_oom && server.maxmemory) {
+        if (!(script_flags & SCRIPT_FLAG_ALLOW_OOM) && server.script_oom && server.maxmemory) { /* unreachable */
```
many of the checks and errors here and below are now dead code.
shall we delete them?

oranagra replied:
the dead error checks are these (in non `SCRIPT_FLAG_EVAL_COMPAT_MODE`):
* `MASTERDOWN`
* `Can not run script with write flag on readonly replica`
* `MISCONF`
* `shared.noreplicaserr`
* `OOM`

the live ones are:
* `Can not run script on cluster, 'no-cluster' flag is set.`
* `Can not execute a script with write flag using *_ro command.`
* `shared.masterdownerr` (the one in the `SCRIPT_FLAG_EVAL_COMPAT_MODE` else)

i suppose that if we can't make them reachable, we should either remove them or panic?

MeirShpilraien replied:
I vote for panic on the deadcode

oranagra replied:
We realized that some of these are reachable from RM_Call.
Usually, when a module calls a redis command via RM_Call, it skips all the checks (Like OOM) and allows the commands to run anyway, i.e. even write commands despite the OOM state.

but when it calls EVAL, the script engine itself would have always used to do the checks when executing the individual commands the script calls, and i suppose we wanna keep that?
so for scripts with flags, it makes sense that these flags would still be tested before the script starts.

that means that we should either keep these checks in `scriptPrepareForRun`, or we should have RM_Call extract the script flags and act accordingly.
i'll add tests for this and try to decide which of these i prefer.

oranagra replied:
added tests for these in my last commit (misc.tcl)

oranagra commented on src/script.c:
```diff
@@ -118,18 +129,18 @@ int scriptPrepareForRun(scriptRunCtx *run_ctx, client *engine_client, client *ca
 
     if (!(script_flags & SCRIPT_FLAG_EVAL_COMPAT_MODE)) {
-        if ((script_flags & SCRIPT_FLAG_NO_CLUSTER) && server.cluster_enabled) {
+        if ((script_flags & SCRIPT_FLAG_NO_CLUSTER) && server.cluster_enabled) { /* unreachable */
```
ignore, i probably didn't run the whole test suite.

MeirShpilraien reviewed (commented):
:+1: LGTM
We should make sure to mentioned all the breaking changes on the release notes (MAY_REPLICATE command which will now be blocked on RO mode..).
Also we should check if there is any docs that need to be updated with respect to those changes.

soloestoy commented:
Mostly LGTM.

But I still feel uncomfortable that when I use `eval_ro` with shebang I have to add `no-writes` flags, I think `eval_ro` with shebang should run with `no-writes` and `allow-oom` automatically, just like what function `evalGetCommandFlags` does:
```c
int evalGetCommandFlags(client *c, uint64_t *flags) {
    char funcname[43];
    int evalsha = c->cmd->proc == evalShaCommand || c->cmd->proc == evalShaRoCommand;
    int ro_cmd = c->cmd->proc == evalRoCommand || c->cmd->proc == evalShaRoCommand;
    ...
    if (ro_cmd)
        script_flags |= SCRIPT_FLAG_NO_WRITES;
    *flags = scriptFlagsToCmdFlags(script_flags);
    return C_OK;
}
```

soloestoy commented:
BTW, now the may-replicate commands are considered write commands in scripts, so we don't need `scriptVerifyMayReplicate` anymore.

```c
static int scriptVerifyMayReplicate(scriptRunCtx *run_ctx, char **err) {
    if (run_ctx->c->cmd->flags & CMD_MAY_REPLICATE &&
        server.client_pause_type == CLIENT_PAUSE_WRITE) {
        *err = sdsnew("May-replicate commands are not allowed when client pause write.");
        return C_ERR;
    }
    return C_OK;
}

oranagra commented:
> But I still feel uncomfortable that when I use `eval_ro` with shebang I have to add `no-writes` flags, I think `eval_ro` with shebang should run with `no-writes` and `allow-oom` automatically, just like what function `evalGetCommandFlags` does:

Think of functions (not eval), the developer declares if the function is read only, or is allowed to ignore OOM state, the user that calls it doesn't necessarily knows what it does (similar to a module command).

I remind us that the reason we added EVAL_RO was for the purpose of client side routing.
one of the main purposes of this PR (second paragraph at the top comment) was for EVAL to be just as good as EVAL_RO (for client pause) for scripts that declared proper flags.

```c
    if (ro_cmd)
        script_flags |= SCRIPT_FLAG_NO_WRITES;
```

i don't know why i added that. it doesn't seem that it's needed for any of my tests, it'll just allow the code to pass `processCommand` and fail in `scriptPrepareForRun`

> BTW, now the may-replicate commands are considered write commands in scripts, so we don't need `scriptVerifyMayReplicate` anymore.

right, i'll delete it.

oranagra commented:
@MeirShpilraien i did list the breaking changes at the top (i invite you to review it and comment if it's unclear), and i'll mention them in the release notes.
i also prepared a doc PR (listed in the top comment), please review that one as well.

oranagra commented:
all, while working on this, and testing OOM errors from EVAL that's called by RM_Call, i found another bug.
the code in `processCommand` that sets `server.script_oom` was trying to avoid changing it while a script a busy script yields, but it meant that EVAL that's called by RM_Call can get executed with an outdated flag, i'll update the top comment to mention that bug too.

@soloestoy please have a look at https://github.com/redis/redis/pull/10744/commits/3bc867f38f3d0d0310d1ec880b91bc5001238769

oranagra commented:
sorry for the mess. moved the `script_oom` bugfix to #10786 where fix other related issues and add a new RM_Call flag that needs it.

oranagra commented:
It's complicated to separate these various fixes since they're all triggered by extra coverage tests, so the tests are all conflicting between these PRs, but i wanna try to have the PRs small with clear topic and description

madolson commented on src/server.c:
```diff
@@ -3615,19 +3615,45 @@ int processCommand(client *c) {
+     * it, in case it declared them. Note this is just an attempt, we don't yet
+     * know the script command is well formed.*/
+    uint64_t cmd_flags = c->cmd->flags;
+    int script_call = 0;
```
For consistency, this can be `is_script_command`.

oranagra replied:
thanks, but in #10786 i completely drop this variable.

madolson commented on src/server.c:
```diff
@@ -3615,19 +3615,45 @@ int processCommand(client *c) {
+        else
+            res = evalGetCommandFlags(c, &fl);
+        if (res == C_OK) {
+            /* If the script declared flags, clear the ones from the command,
+             * specifically STALE, and use the ones it declared.
+             * In addition the MAY_REPLICATE flag is set for these commands, but
+             * if we have flags we know if it's gonna do any writes or not. */
```
This comment is a little confusing. It calls out STALE specifically but not CMD_DENYOOM and CMD_WRITE, which I presume is because DENYOOM and WRITE aren't on EVAL, but it makes sense to remove them because it's defensive.

A alternative might be easy to follow by changing the `fcallGetCommandFlags` so that it just returns the "known" flags for the command, and it internally has the logic to remove flags. So that the code here is just
```
        if (c->cmd->proc == fcallCommand || c->cmd->proc == fcallroCommand)
            cmd_flags = fcallGetCommandFlags(c, cmd_flags);
        else
            cmd_flags = evalGetCommandFlags(c, cmd_flags);
```

The relevant logic can be moved to scriptFlagsToCmdFlags, which will make it easier to follow since the code is right there.
```
uint64_t scriptFlagsToCmdFlags(uint64_t initial_flags, int script_flags) {
    /* If the script declared flags, clear the ones from the command and use the ones it declared.
    initial_flags &= ~(CMD_STALE | CMD_DENYOOM | CMD_WRITE);
    /* NO_WRITES implies ALLOW_OOM */
    if (!(script_flags & (SCRIPT_FLAG_ALLOW_OOM | SCRIPT_FLAG_NO_WRITES)))
        initial_flags |= CMD_DENYOOM;
    if (!(script_flags & SCRIPT_FLAG_NO_WRITES))
        initial_flags |= CMD_WRITE;
    if (script_flags & SCRIPT_FLAG_ALLOW_STALE)
        initial_flags |= CMD_STALE;
    /* In addition the MAY_REPLICATE flag is set for script commands, but
     * if we have flags we know if it's gonna do any writes or not. */
    initial_flags &= ~CMD_MAY_REPLICATE;
    return initial_flags;
}
```

oranagra replied:
i considered this approach and abandoned it, but i don't remember why.. maybe because of MAY_REPLICATE.
i'll take it and see what happens.

madolson commented on tests/unit/pause.tcl:
```diff
@@ -78,20 +74,121 @@ start_server {tags {"pause network"}} {
+        $rd2 close
+    }
+
+    test "Test may-replicate commands are rejected in ro script RO scripts" {
```
```suggestion
    test "Test may-replicate commands are rejected in RO scripts" {
```

madolson reviewed (approved):
Logic all makes sense.

soloestoy commented:
> Think of functions (not eval), the developer declares if the function is read only, or is allowed to ignore OOM state, the user that calls it doesn't necessarily knows what it does (similar to a module command).

I know the flags are script's attributes not the command's, but `eval` and `eval_ro` have a different with functions, they can call and modify the script's body, that makes `eval` and `eval_ro` more complex. It's OK to me, but I'm afraid some users may misunderstand it.

Another question I wanna discuss, now when write paused, `eval_ro` and `fcall_ro` will be blocked if using them to call a script without `no-writes` flag, do you thinks it's better to return an error instead of block them?

oranagra commented:
> I know the flags are script's attributes not the command's, but eval and eval_ro have a different with functions, they can call and modify the script's body, that makes eval and eval_ro more complex. It's OK to me, I don't insist on changing it, but I'm afraid some users may misunderstand it.

are you also suggesting a different default behavior for EVAL_RO vs EVALSHA_RO?
i.e. that since EVAL_RO has the script embedded in the RO command it'll have default of `no-writes` and for EVALSHA_RO it'll be similar to FCALL_RO (same defaults as normal FCALL)?

i think we should keep these _RO commands the same as the non-RO commands, and just say that they're for client routing and ACL.

> Another question I wanna discuss, now when write paused, eval_ro and fcall_ro will be blocked if using them to call a script without no-writes flag, do you thinks it's better to return an error instead of block them?

in theory, your suggestion would make a better experience (return error right away, rather than block and return error when unblocked). but in practice, it's a programming mistake that will be discovered by the programmer when testing the script (on normal state), long before ever getting to run it in CLIENT PAUSE state, so it doesn't make any difference, and i rather not complicate the redis code for that (whatever simpler is good in my book)

soloestoy commented:
> are you also suggesting a different default behavior for EVAL_RO vs EVALSHA_RO?
i.e. that since EVAL_RO has the script embedded in the RO command it'll have default of no-writes and for EVALSHA_RO it'll be similar to FCALL_RO (same defaults as normal FCALL)?

nope, I don't want make it too much complex, just keep it
